File: app/src/main/java/com/example/fsudouest/blablafit/MyProfileFragment.java
package com.example.fsudouest.blablafit;


import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.core.content.ContextCompat;
import androidx.appcompat.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.firebase.ui.auth.AuthUI;
import com.google.android.gms.tasks.Continuation;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.UserProfileChangeRequest;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;

import static android.app.Activity.RESULT_OK;


/**
 * A simple {@link Fragment} subclass.
 */
public class MyProfileFragment extends Fragment {

    private static final int RC_PHOTO_PICKER = 2;

    private FirebaseStorage mFirebaseStorage;
    private StorageReference mProfilePhotosStorageReference;
    FirebaseUser user;
    ImageView profile_pic;
    public MyProfileFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View rootview = inflater.inflate(R.layout.fragment_my_profile, container, false);

        user =FirebaseAuth.getInstance().getCurrentUser();
        mFirebaseStorage = FirebaseStorage.getInstance();
        mProfilePhotosStorageReference = mFirebaseStorage.getReference().child("profile_pictures");

        TextView pseudo = rootview.findViewById(R.id.pseudo);
        TextView nom = rootview.findViewById(R.id.nom);
        TextView email = rootview.findViewById(R.id.email);
        profile_pic = rootview.findViewById(R.id.profile_pic_image_view);

        pseudo.setText(user.getDisplayName());
        nom.setText("Nom: "+ user.getDisplayName());
        //prenom.setText("Prénom: "+preferences.getString("prénom",null));
        email.setText("Email: "+user.getEmail());
        if(user.getPhotoUrl()!=null){
            Glide.with(getContext())
                    .load(user.getPhotoUrl().toString())
                    .into(profile_pic);
        }

        profile_pic.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                createPhotoUpdateDialog();
            }
        });

        Button deconnexion = rootview.findViewById(R.id.deco_button);
        deconnexion.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                signOut();
            }
        });
        return rootview;
    }


    private void chooseImageFromGallery() {
        Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
        intent.setType("image/*");
        intent.putExtra(Intent.EXTRA_LOCAL_ONLY, true);
        startActivityForResult(Intent.createChooser(intent, "Complete action using"), RC_PHOTO_PICKER);
    }

    private void createPhotoUpdateDialog(){
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());

        builder.setTitle(getString(R.string.update_profile_dialog_title));
        builder.setMessage(getString(R.string.update_photo_dialog_message));
        // Add the button
        builder.setPositiveButton("OK", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int id) {
                // User clicked OK button
                chooseImageFromGallery();
            }
        });
        builder.setNegativeButton(getString(R.string.cancel), new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                //
            }
        });
        // Create the AlertDialog
        AlertDialog dialog = builder.create();
        dialog.show();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == RC_PHOTO_PICKER && resultCode == RESULT_OK) {
            Uri selectedImageUri = data.getData();
            profile_pic.setImageURI(selectedImageUri);
            // Get a reference to store file at profile_pictures/<FILENAME>
            final StorageReference photoRef = mProfilePhotosStorageReference.child(selectedImageUri.getLastPathSegment());
            photoRef.putFile(selectedImageUri).continueWithTask(new Continuation<UploadTask.TaskSnapshot, Task<Uri>>() {
                @Override
                public Task<Uri> then(@NonNull Task<UploadTask.TaskSnapshot> task) throws Exception {
                    if (!task.isSuccessful()) {
                        throw task.getException();
                    }
                    // Continue with the task to get the download URL
                    return photoRef.getDownloadUrl();
                }
            }).addOnCompleteListener(new OnCompleteListener<Uri>() {
                @Override
                public void onComplete(@NonNull Task<Uri> task) {
                    if (task.isSuccessful()) {
                        Uri downloadUri = task.getResult();
                        UserProfileChangeRequest photoUpdate = new UserProfileChangeRequest.Builder()
                                .setPhotoUri(downloadUri)
                                .build();
                        user.updateProfile(photoUpdate);
                    } else {
                        Toast.makeText(getActivity(),"Upload Failed - please try again",Toast.LENGTH_SHORT).show();
                    }
                }
            });

        }
    }

    private void signOut() {
        AuthUI.getInstance().signOut(getActivity());
    }

}

File: app/src/main/java/com/example/fsudouest/blablafit/Ui/Fragments/MyProfileFragment.kt
package com.example.fsudouest.blablafit.Ui.Fragments


import android.content.Intent
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.appcompat.app.AlertDialog
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast

import com.bumptech.glide.Glide
import com.firebase.ui.auth.AuthUI
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.UserProfileChangeRequest
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference

import android.app.Activity.RESULT_OK
import androidx.databinding.DataBindingUtil
import com.example.fsudouest.blablafit.R
import com.example.fsudouest.blablafit.databinding.FragmentMyProfileBinding
import com.example.fsudouest.blablafit.di.Injectable
import com.example.fsudouest.blablafit.model.User
import javax.inject.Inject


/**
 * A simple [Fragment] subclass.
 */
class MyProfileFragment : Fragment(), Injectable {

    @Inject
    lateinit var mFirebaseStorage: FirebaseStorage

    @Inject
    lateinit var mFirebaseAuth: FirebaseAuth


    private var mProfilePhotosStorageReference: StorageReference? = null
    private var firebaseUser: FirebaseUser? = null
    private lateinit var profile_pic: ImageView


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val binding: FragmentMyProfileBinding = DataBindingUtil.inflate(inflater,R.layout.fragment_my_profile, container, false)


        firebaseUser = mFirebaseAuth.currentUser

        mProfilePhotosStorageReference = mFirebaseStorage.reference.child("profile_pictures")

        binding.user = User(firebaseUser?.displayName!!,firebaseUser?.email!!)
        profile_pic = binding.profilePicImageView


        if (firebaseUser!!.photoUrl != null) {
            Glide.with(context!!)
                    .load(firebaseUser!!.photoUrl!!.toString())
                    .into(profile_pic)
        }

        profile_pic.setOnClickListener { createPhotoUpdateDialog() }

        binding.decoButton.setOnClickListener { signOut() }
        return binding.root
    }


    private fun chooseImageFromGallery() {
        val intent = Intent(Intent.ACTION_GET_CONTENT)
        intent.type = "image/*"
        intent.putExtra(Intent.EXTRA_LOCAL_ONLY, true)
        startActivityForResult(Intent.createChooser(intent, "Complete action using"), RC_PHOTO_PICKER)
    }

    private fun createPhotoUpdateDialog() {
        val builder = AlertDialog.Builder(activity!!)

        builder.setTitle(getString(R.string.update_profile_dialog_title))
        builder.setMessage(getString(R.string.update_photo_dialog_message))
        // Add the button
        builder.setPositiveButton("OK") { dialog, id ->
            // User clicked OK button
            chooseImageFromGallery()
        }
        builder.setNegativeButton(getString(R.string.cancel)) { dialogInterface, i ->
            //
        }
        // Create the AlertDialog
        val dialog = builder.create()
        dialog.show()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == RC_PHOTO_PICKER && resultCode == RESULT_OK) {
            val selectedImageUri = data!!.data
            profile_pic.setImageURI(selectedImageUri)
            // Get a reference to store file at profile_pictures/<FILENAME>
            val photoRef = mProfilePhotosStorageReference!!.child(selectedImageUri!!.lastPathSegment!!)
            photoRef.putFile(selectedImageUri).continueWithTask { task ->
                if (!task.isSuccessful) {
                    throw task.exception!!.fillInStackTrace()
                }
                // Continue with the task to get the download URL
                photoRef.downloadUrl
            }.addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    val downloadUri = task.result
                    val photoUpdate = UserProfileChangeRequest.Builder()
                            .setPhotoUri(downloadUri)
                            .build()
                    firebaseUser!!.updateProfile(photoUpdate)
                } else {
                    Toast.makeText(activity, "Upload Failed - please try again", Toast.LENGTH_SHORT).show()
                }
            }

        }
    }

    private fun signOut() {
        AuthUI.getInstance().signOut(activity!!)
    }

    companion object {

        private val RC_PHOTO_PICKER = 2
    }

}
